service/src/test/java/bio/terra/workspace/app/configuration/external/controller/ResourceApiControllerConnectedTest.java

```diff
@@ -0,0 +1,320 @@
+import static bio.terra.workspace.common.utils.MockMvcUtils.RESOURCE_PROPERTIES_V1_PATH_FORMAT;
+import static bio.terra.workspace.common.utils.MockMvcUtils.addAuth;
+import static bio.terra.workspace.common.utils.MockMvcUtils.addJsonContentType;
+import static bio.terra.workspace.common.utils.MockMvcUtils.createDefaultBigQueryDataset;
+import static bio.terra.workspace.common.utils.MockMvcUtils.deleteWorkspace;
+import static bio.terra.workspace.common.utils.MockMvcUtils.getBigQueryDataset;
+import static bio.terra.workspace.common.utils.MockMvcUtils.grantRole;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+import bio.terra.workspace.common.BaseConnectedTest;
+import bio.terra.workspace.connected.UserAccessUtils;
+import bio.terra.workspace.connected.WorkspaceConnectedTestUtils;
+import bio.terra.workspace.generated.model.ApiCreatedControlledGcpBigQueryDataset;
+import bio.terra.workspace.generated.model.ApiGcpBigQueryDatasetResource;
+import bio.terra.workspace.generated.model.ApiProperties;
+import bio.terra.workspace.generated.model.ApiPropertyKeys;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.model.WsmIamRole;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.apache.http.HttpStatus;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+@AutoConfigureMockMvc
+public class ResourceApiControllerConnectedTest extends BaseConnectedTest {
+
+  @Autowired MockMvc mockMvc;
+  @Autowired ObjectMapper objectMapper;
+  @Autowired UserAccessUtils userAccessUtils;
+  @Autowired WorkspaceConnectedTestUtils connectedTestUtils;
+
+  private UUID workspaceId;
+
+  @BeforeEach
+  public void setUp() throws Exception {
+    workspaceId =
+        connectedTestUtils
+            .createWorkspaceWithGcpContext(userAccessUtils.defaultUserAuthRequest())
+            .getWorkspaceId();
+  }
+
+  @AfterEach
+  public void cleanup() throws Exception {
+    deleteWorkspace(workspaceId, mockMvc, userAccessUtils.defaultUserAuthRequest());
+  }
+
+  @Nested
+  class UpdateResourceProperties {
```

Comment:
For long tests, it helps the reader if you add comments. One comment for each "step".

- They can skim the test by adding comments
- I had to dig to find out what properties original resource had


```
     // Create resource with property foo=bar
     ApiCreatedControlledGcpBigQueryDataset resource =
          createDefaultBigQueryDataset(
              mockMvc, objectMapper, workspaceId, userAccessUtils.defaultUserAuthRequest());

      // Add two more properties with keys terra_workspace_folder_id, data_type
      updateResourcePropertiesExpectCode(
          workspaceId, resourceId, newProperties, HttpStatus.SC_NO_CONTENT);

      // Assert resource has three properties
      ApiGcpBigQueryDatasetResource updatedResource =
          getBigQueryDataset(
              mockMvc,
              objectMapper,
              workspaceId,
              resourceId,
              userAccessUtils.defaultUserAuthRequest());
      Map<String, String> expectedProperties = HashMap.of(...);
      assertEquals(
          expectedProperties,
          convertApiPropertyToMap(updatedResource.getMetadata().getProperties()));

      // Change property terra_workspace_folder_id to new UUID
      var newFolderId = UUID.randomUUID();
      updateResourcePropertiesExpectCode(
          workspaceId,
          resourceId,
          Map.of(folderIdKey, newFolderId.toString()),
          HttpStatus.SC_NO_CONTENT);

      // Assert terra_workspace_folder_id has new UUID
      ApiGcpBigQueryDatasetResource updatedResource2 =
          getBigQueryDataset(
              mockMvc,
              objectMapper,
              workspaceId,
              resourceId,
              userAccessUtils.defaultUserAuthRequest());
      assertEquals(
          newFolderId.toString(),
          convertApiPropertyToMap(updatedResource2.getMetadata().getProperties()).get(folderIdKey));
```

I also simplified expectedProperties. Before: first we call GET to get original properties. After: Don't call GET, just hard-code the three expected properties.

Reply:
added more comments.

btw you can't create a resource with property foo and bar. properties can only be added through the updateResourceProperties end point. That's why in the deleteResourceProperties test, I have to always call updateResourceProperties in the setup before testing delete.

Reply:
I expect you will be adding the ability to specify properties on resource create, also, yes?

Reply:
ooh.. i falsely remember that properties are removed from workspace creation. Yeah I will have a follow-on PR for that